### PR TITLE
IoTHubBuddy

### PR DESCRIPTION
**Pull request by DivyaMahadevan**
Adding the basic app to the repo. Authentication still doesn't work in IoTHubBuddy, but there are instructions in MainPage.xaml.cs on how to skip authentication step and test the rest of the app.

### CODE REVIEW

File: IoTSender/IoTSender/App.xaml.cs [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+
+        /// <summary>
+        /// Invoked when the application is launched normally by the end user.  Other entry points
+        /// will be used such as when the application is launched to open a specific file.
+        /// </summary>
+        /// <param name="e">Details about the launch request and process.</param>
+        protected override void OnLaunched(LaunchActivatedEventArgs e)
+        {
+#if DEBUG
+            if (System.Diagnostics.Debugger.IsAttached)
```
Comment (arturl) [inline IoTSender/IoTSender/App.xaml.cs]: this is safe to remove -- it produces annoying frame counters that you don't need.

File: IoTHubBuddy/IoTViewer/AccountManager.cs [hunk 2]
```diff
@@ -0,0 +1,362 @@
...
+           AccountsSettingsPaneCommandsRequestedEventArgs e, MainPage page)
+        {
+            mp = page;
+            // In order to make async calls within this callback, the deferral object is needed
+            AccountsSettingsPaneEventDeferral deferral = e.GetDeferral();
+            CreateLocalDataContainers();
+            await AddWebAccountProvidersToPane(e);
+            await AddWebAccountsToPane(e);
+
+            deferral.Complete();
```
Comment (arturl) [inline IoTHubBuddy/IoTViewer/AccountManager.cs:55]: for my education: why is this necessary?

File: IoTHubBuddy/IoTViewer/AccountManager.cs [hunk 3]
```diff
@@ -0,0 +1,362 @@
...
+            foreach (WebAccount account in accounts)
+            {
+                WebAccountCommand command = new WebAccountCommand(account, WebAccountInvoked, SupportedWebAccountActions.Remove);
+                e.WebAccountCommands.Add(command);
+            }
+
+        }
+        /// <summary>
+        /// Get a list of stored accounts
+        /// NOTE: the storage model is old, and will be changed soon
```
Comment (arturl) [inline IoTHubBuddy/IoTViewer/AccountManager.cs:90]: old comment?

File: IoTHubBuddy/IoTViewer/AccountManager.cs [hunk 4]
```diff
@@ -0,0 +1,362 @@
...
+
+            foreach (Object value in AccountListContainer.Containers[ProviderIdSubContainer].Values.Keys)
+            {
+                String accountID = value as String;
+                //AccountListContainer.Containers[ProviderIdSubContainer].Values.Remove(accountID);
+                String providerID = AccountListContainer.Containers[ProviderIdSubContainer].Values[accountID] as String;
+                String authority = AccountListContainer.Containers[AuthoritySubContainer].Values[accountID] as String;
+                if(authority == null)
+                {
+                    authority = "organizations";
```
Comment (arturl) [inline IoTHubBuddy/IoTViewer/AccountManager.cs:107]: why organizations and not consumers?

File: IoTHubBuddy/IoTViewer/AccountManager.cs [hunk 5]
```diff
@@ -0,0 +1,362 @@
...
+        /// <param name="client"></param>
+        /// <returns></returns>
+        public static async Task AuthenticateWithRequestToken(WebAccountProvider provider, string scope, string client)
+        {
+            WebTokenRequest request = new WebTokenRequest(provider, scope, client);
+            Debug.WriteLine("provider: " + provider.Id + " scope: " + scope + " client: " + client);
+            if (provider.Id == "https://login.microsoft.com")
+            {
+                request.Properties.Add("resource", "https://management.azure.com/");
+            }
```
Comment (arturl) [inline IoTHubBuddy/IoTViewer/AccountManager.cs]: and if provider.Id is something else, then what?

File: IoTHubBuddy/IoTViewer/AccountManager.cs [hunk 6]
```diff
@@ -0,0 +1,362 @@
...
+            
+
+        }
+
+        /// <summary>
+        /// store web account in local storage
+        /// NOTE: this will likely change soon
+        /// </summary>
+        /// <param name="account"></param>
+        private static async void StoreWebAccount(WebAccount account)
```
Comment (arturl) [inline IoTHubBuddy/IoTViewer/AccountManager.cs]: why is this function async? I don't see any awaits in it
also, please audit your code for async void functions -- these are bad (event handler have to be like that, but not other methods)

File: IoTHubBuddy/IoTViewer/AccountManager.cs [hunk 7]
```diff
@@ -0,0 +1,362 @@
...
+        /// <returns></returns>
+        public static async Task<string> GetTokenSilentlyAsync()
+        {
+            string providerId = ApplicationData.Current.LocalSettings.Values["CurrentUserProviderId"]?.ToString();
+            string accountId = ApplicationData.Current.LocalSettings.Values["CurrentUserId"]?.ToString();
+            string authority = ApplicationData.Current.LocalSettings.Values["CurrentAuthority"]?.ToString();
+
+            if (null == providerId || null == accountId)
+            {
+                return null;
```
Comment (arturl) [inline IoTHubBuddy/IoTViewer/AccountManager.cs]: Avoid 'return null' -- this just makes callers life miserable ("billion dollar mistake")

Applies everywhere :-)

File: IoTHubBuddy/IoTViewer/AccountManager.cs [hunk 8]
```diff
@@ -0,0 +1,362 @@
...
+        /// <returns></returns>
+        public static async Task SignOutAccountAsync(WebAccount account)
+        {
+            ApplicationData.Current.LocalSettings.Values.Remove("CurrentUserProviderId");
+            ApplicationData.Current.LocalSettings.Values.Remove("CurrentUserId");
+            ApplicationData.Current.LocalSettings.Values.Remove("CurrentAuthority");
+            ApplicationDataContainer accountsContainer = ApplicationData.Current.LocalSettings.Containers[AccountsContainer];
+            if(account.UserName=="")
+            {
+                accountsContainer.Containers[ProviderIdSubContainer].Values.Remove(account.Id);
```
Comment (arturl) [inline IoTHubBuddy/IoTViewer/AccountManager.cs:311]: hm... this sounds suspect. what is stored in the hashtable -- user named or ids?

File: IoTHubBuddy/IoTViewer/AccountManager.cs [hunk 9]
```diff
@@ -0,0 +1,362 @@
...
+            
+        }
+        /// <summary>
+        /// attempt to login silently without prompting user for credentials
+        /// </summary>
+        /// <returns></returns>
+        public static async Task<string> LoginSilently()
+        {
+            string token = await GetTokenSilentlyAsync();
+            if(token != null)
```
Comment (arturl) [inline IoTHubBuddy/IoTViewer/AccountManager.cs]: Useless. Just return token.

File: IoTHubBuddy/IoTViewer/AccountManager.cs [hunk 10]
```diff
@@ -0,0 +1,362 @@
...
+        public static async Task LogoffAndRemoveAllAccounts()
+        {
+            List<WebAccount> accounts = await GetAllAccounts();
+
+            foreach (WebAccount account in accounts)
+            {
+                await SignOutAccountAsync(account);
+            }
+
+            accounts = null;
```
Comment (arturl) [inline IoTHubBuddy/IoTViewer/AccountManager.cs]: dead store

File: IoTHubBuddy/IoTViewer/DevicePage.xaml.cs [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+            
+
+        }
+        private void ItemSelected(object sender, ItemClickEventArgs e)
+        {
+            string device = e.ClickedItem.ToString();
+            data.DeviceName = device;;
+            this.Frame.Navigate(typeof(MapPage), data);
+        }
+        private bool checkValidity(IoTAccountData account)
```
Comment (arturl) [inline IoTHubBuddy/IoTViewer/DevicePage.xaml.cs]: nitpick on styling -- showError and hideErrors are in camelCase, other names are PascalCase.

File: IoTHubBuddy/IoTViewer/HubPage.xaml.cs [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+                if(hubs.Count == 0)
+                {
+                    showError("You have no IoTHubs registered under this subscription.");
+                } else
+                {
+                    hideErrors();
+                    foreach (EventHubData h in hubs)
+                    {
+                        HubList.Items.Add(h.HubName);
+                    }
```
Comment (arturl) [inline IoTHubBuddy/IoTViewer/HubPage.xaml.cs:50]: how about: 
`HubList.Items.AddRange(hubs.Select(_ => _.HubName));` 
?

(Avoiding loops, my other pet peeve. Only nulls are worse than loops :-))

File: IoTHubBuddy/IoTViewer/HubPage.xaml.cs [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+            
+            data.HubName = hub;
+            EventHubData eventhub = EventSelected(hub);
+            eventhub.HubName += ".azure-devices.net";
+            data.EventHubInfo = eventhub;
+            this.Frame.Navigate(typeof(DevicePage), data);
+        }
+        private EventHubData EventSelected(string hubName)
+        {
+            foreach(EventHubData e in hubs)
```
Comment (arturl) [inline IoTHubBuddy/IoTViewer/HubPage.xaml.cs]: loop!

File: IoTHubBuddy/IoTViewer/HubPage.xaml.cs [hunk 14]
```diff
@@ -0,0 +1,125 @@
...
+            {
+                if(e.HubName == hubName)
+                {
+                    return e;
+                }
+            }
+            return null;
```
Comment (arturl) [inline IoTHubBuddy/IoTViewer/HubPage.xaml.cs]: null!

File: IoTHubBuddy/IoTViewer/IoTDataManager.cs [hunk 15]
```diff
@@ -0,0 +1,228 @@
...
+        private const string AzureResourceUri = "https://management.azure.com/";
+
+        /// <summary>
+        /// check validity for returned json object. If it doesn't contain a value, then there was an error in gathering the data
+        /// Error parsing/handling will be implemented later
+        /// </summary>
+        /// <param name="jo"></param>
+        /// <returns></returns>
+        private static bool checkJsonObject(JsonObject jo)
+        {
```
Comment (arturl) [inline IoTHubBuddy/IoTViewer/IoTDataManager.cs:29]: how about: `return jo.ContainsKey("value");`

File: IoTHubBuddy/IoTViewer/IoTDataManager.cs [hunk 16]
```diff
@@ -0,0 +1,228 @@
...
+        /// sends a GET resquest to azure management REST API
+        /// </summary>
+        /// <param name="token"></param>
+        /// <param name="relative"></param>
+        /// <returns></returns>
+        public static async Task<JsonObject> GetIoTData(string token, string relative)
+        {
+            if(token == null)
+            {
+                return null;
```
Comment (arturl) [inline IoTHubBuddy/IoTViewer/IoTDataManager.cs]: oops! 
assert that `token` is not null

File: IoTHubBuddy/IoTViewer/IoTDataManager.cs [hunk 17]
```diff
@@ -0,0 +1,228 @@
...
+                var restApi = new Uri(AzureResourceUri + relative);
+                var infoResult = await client.GetAsync(restApi);
+                string content = await infoResult.Content.ReadAsStringAsync();
+                var jsonObject = JsonObject.Parse(content);
+                if(checkJsonObject(jsonObject))
+                {
+                   return jsonObject;
+                } else
+                {
+                    return null;
```
Comment (arturl) [inline IoTHubBuddy/IoTViewer/IoTDataManager.cs]: I will stop complaining about nulls now, but please audit the code for `return null` and see if you can avoid it.

File: IoTHubBuddy/IoTViewer/IoTDataManager.cs [hunk 18]
```diff
@@ -0,0 +1,228 @@
...
+        }
+        /// <summary>
+        /// get a list of resource group for a specific azure subscription
+        /// </summary>
+        /// <param name="subId"></param>
+        /// <returns></returns>
+        public static async Task<ICollection<string>> GetResourceGroups(string subId)
+        {
+            List<string> groupNames = new List<string>();
+            var relative = "subscriptions/" + subId + "/resourcegroups?api-version=2015-01-01";
```
Comment (arturl) [inline IoTHubBuddy/IoTViewer/IoTDataManager.cs]: Can we factor out the api version into a variable, for consistency and so it can be bumped in one place

File: IoTHubBuddy/IoTViewer/IoTDataManager.cs [hunk 19]
```diff
@@ -0,0 +1,228 @@
...
+                foreach (var hub in hubs)
+                {
+                    var events = hub.GetObject().GetNamedValue("properties").GetObject().GetNamedValue("eventHubEndpoints").GetObject().GetNamedValue("events").GetObject();
+                    string idContent = hub.GetObject().GetNamedValue("name").GetString();
+                    var partIds = events.GetNamedValue("partitionIds").GetArray();
+                    List<string> ids = new List<string>();
+                    foreach (var id in partIds)
+                    {
+                        ids.Add(id.GetString());
+                    }
```
Comment (arturl) [inline IoTHubBuddy/IoTViewer/IoTDataManager.cs:186]: consider: `var ids = partIds.Select(_ => _.GetString()).ToList();`

File: IoTHubBuddy/IoTViewer/MainPage.xaml.cs [hunk 20]
```diff
@@ -0,0 +1,131 @@
...
+        /// sign in silently first. If that fails, then bring up gui
+        /// </summary>
+        /// <param name="sender"></param>
+        /// <param name="e"></param>
+        private async void On_Loaded(object sender, RoutedEventArgs e)
+        {
+            List<string> partitionIds = new List<string>();
+            partitionIds.Add("0");
+            partitionIds.Add("1");
+            string entity = "iothub-ehub-hub-dimaha-42362-31f5229b77"; //should start with "iothub-ehub-<hub name>-<random numbers>"
```
Comment (arturl) [inline IoTHubBuddy/IoTViewer/MainPage.xaml.cs]: remember to not check this in. 
any customizable code should be in one place, so it's easy to find.

File: IoTHubBuddy/IoTViewer/MessageManager.cs [hunk 21]
```diff
@@ -0,0 +1,80 @@
...
+        }
+        /// <summary>
+        /// parse message, check validity of message format, display coordinates or error message accordingly
+        /// </summary>
+        /// <param name="msg"></param>
+        public void parseMessage(string msg)
+        {
+            JObject jsonMsg = JObject.Parse(msg);
+            Debug.WriteLine(jsonMsg);
+            if (validateMessage(jsonMsg))
```
Comment (arturl) [inline IoTHubBuddy/IoTViewer/MessageManager.cs]: I suggest doing validation in place to avoid extra lookups ( `var a = x["y"]; if(!a) a["z"] ...` )

File: IoTHubBuddy/IoTViewer/MessageManager.cs [hunk 22]
```diff
@@ -0,0 +1,80 @@
...
+            {
+                string type = jsonMsg["message"]["type"].ToString();
+                if (type == "coordinates")
```
Comment (arturl) [inline IoTHubBuddy/IoTViewer/MessageManager.cs]: and if it's not?

File: IoTHubBuddyClient/IoTSender/Interfaces/Http/WebHelper.cs [hunk 23]
```diff
@@ -0,0 +1,202 @@
...
+                        await resp.WriteAsync(headerArray, 0, headerArray.Length);
+                        await fs.CopyToAsync(resp);
+                    }
+                }
+                catch (FileNotFoundException ex)
+                {
+                    exists = false;
+
+                    // Log telemetry event about this exception
+                    var events = new Dictionary<string, string> { { "WebHelper", ex.Message } };
```
Comment (arturl) [inline IoTHubBuddyClient/IoTSender/Interfaces/Http/WebHelper.cs]: dead store?

File: IoTHubBuddyClient/IoTSender/Interfaces/Http/WebHelper.cs [hunk 24]
```diff
@@ -0,0 +1,202 @@
...
+                }
+
+                if (!exists)
```
Comment (arturl) [inline IoTHubBuddyClient/IoTSender/Interfaces/Http/WebHelper.cs]: why not do this inside catch, and avoid the `exists` variable?

File: IoTHubBuddy/IoTViewer/AccountManager.cs (hunk 2 above)
Comment (DivyaMahadevan) [inline IoTHubBuddy/IoTViewer/AccountManager.cs:55]: the deferral is there to delay showing the AccountsSettingsPane until I'm finished building it

File: IoTHubBuddy/IoTViewer/AccountManager.cs (hunk 3 above)
Comment (DivyaMahadevan) [inline IoTHubBuddy/IoTViewer/AccountManager.cs:90]: Nope. The storage system in place right now is from the WAM tutorial. If I can get WAM to work, I was planning on looking into the APIs you sent me about storage account info and changing it.

File: IoTHubBuddy/IoTViewer/AccountManager.cs (hunk 8 above)
Comment (DivyaMahadevan) [inline IoTHubBuddy/IoTViewer/AccountManager.cs:311]: this is part of the reason why I want to change the storage layout. The consumer/organization accounts are set up differently so if account.UserName exists, then it is a consumer account and the account is stored with that. Otherwise, the account is stored by its id, and I need to retrieve it that way. I based my storage off the WAM sample (found [here](https://github.com/Microsoft/Windows-universal-samples/tree/master/Samples/WebAccountManagement)) but it's not ideal.

File: IoTHubBuddy/IoTViewer/HubPage.xaml.cs (hunk 12 above)
Comment (DivyaMahadevan) [inline IoTHubBuddy/IoTViewer/HubPage.xaml.cs:50]: unfortunately, ItemCollection does not have an addRange method or the equivalent. I cannot think of another way to iterate over the whole list without using a for loop
